### PR TITLE
ceph-volume: retrieve device data concurrently

### PR DESCRIPTION
**Pull request by jan--f**
Signed-off-by: Jan Fajerski <jfajerski@suse.com>

This implements concurrent device retrieval using futures.
One issue that still needs figuring out with this is that terminal logging from the threads doesn't seem to work. Terminal logging will also likely need synchronisation, since otherwise jumbled log messages are expected.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by pcuzner** [comment]
@jan--f Slightly annoyed with this - given the idea comes from my PR and the other way of doing this is nearly 1s faster - but you're the boss!

File: src/ceph-volume/ceph_volume/util/device.py [hunk 1]
```diff
@@ -30,8 +35,21 @@ class Devices(object):
     def __init__(self, devices=None):
         if not sys_info.devices:
             sys_info.devices = disk.get_devices()
-        self.devices = [Device(k) for k in
-                            sys_info.devices.keys()]
+        self.devices = []
+        with futures.ThreadPoolExecutor() as executor:
```
Comment (pcuzner) [inline src/ceph-volume/ceph_volume/util/device.py:39]: have you considered using max_workers. The default is quite high and in my tests leads to longer runtimes. When I ran with max_workers=4 I was getting consistently shorter runtimes (.5->1s less) - same result as when I ran the threads directly in my PR.

Comment (jan--f) [reply src/ceph-volume/ceph_volume/util/device.py:39]: Yep, can confirm that. 
|                     | max_workers=0 | max_workers=2 | max_workers=4 | max_workers=8 |
|------------------|-----------------------|-----------------------|-----------------------|-----------------------|
|avg runtime in ms over 10 runs |2366.1|2396.3|2189.7|2362.7|

**Comment by jan--f** [comment]
jenkins test ceph-volume tox

**Comment by jan--f** [comment]
jenkins test ceph-volume all

**Review by pcuzner — APPROVED** [review]
LGTM

**Comment by jan--f** [comment]
jenkins test ceph-volume all

**Comment by jan--f** [comment]
hmm the functional tests in centos 8 still run in python2. @guits do you know if a centos 8 rpm install is also still using python2 or is this an artifact of our functional tests?

**Comment by guits** [comment]
jenkins test ceph-volume batch centos8-bluestore-mixed_type

**Comment by guits** [comment]
jenkins test ceph-volume lvm centos8-filestore-dmcrypt

**Comment by guits** [comment]
@jan--f it looks like we enforce python 2.7 if available, see https://github.com/ceph/ceph-build/blob/master/scripts/build_utils.sh#L94-L105
@djgalloway is there a way to make the CI smarter and use the right python version according to the ceph branch being tested?

**Comment by djgalloway** [comment]
> @djgalloway is there a way to make the CI smarter and use the right python version according to the ceph branch being tested?

Just for ceph-volume jobs?

**Comment by guits** [comment]
> > @djgalloway is there a way to make the CI smarter and use the right python version according to the ceph branch being tested?
> 
> Just for ceph-volume jobs?

not especially, maybe other projects as well could be interested in this

**Comment by djgalloway** [comment]
jenkins test ceph-volume all

**Comment by djgalloway** [comment]
> > > @djgalloway is there a way to make the CI smarter and use the right python version according to the ceph branch being tested?
> > 
> > 
> > Just for ceph-volume jobs?
> 
> not especially, maybe other projects as well could be interested in this

Trying this change to build_utils.sh

```
diff --git a/scripts/build_utils.sh b/scripts/build_utils.sh
index 62541041..a61da253 100644
--- a/scripts/build_utils.sh
+++ b/scripts/build_utils.sh
@@ -96,10 +96,12 @@ create_virtualenv () {
     if [ "$(ls -A $path)" ]; then
         echo "Will reuse existing virtual env: $path"
     else
-        if [ $(command -v python2.7) ]; then
+        if [ $(command -v python3) ]; then
+            virtualenv -p python3 $path
+        elif [ $(command -v python2.7) ]; then
             virtualenv -p python2.7 $path
         else
-            virtualenv -p python3 $path
+            virtualenv -p python $path
         fi
     fi
 }
```

**Comment by djgalloway** [comment]
jenkins test ceph-volume all

**Comment by djgalloway** [comment]
jenkins test centos8-filestore-mixed_type

**Comment by djgalloway** [comment]
@guits is CentOS8 as the base operating system a requirement for the ceph-volume tox scenarios to pass?

**Comment by guits** [comment]
> @guits is CentOS8 as the base operating system a requirement for the ceph-volume tox scenarios to pass?

I believe yes but to be honest, I'm unsure. I would need to test.

**Comment by guits** [comment]
@djgalloway so from what I'm understanding, CentOS 8 isn't a requirement for the ceph-volume tox testing.

**Comment by djgalloway** [comment]
> @djgalloway so from what I'm understanding, CentOS 8 isn't a requirement for the ceph-volume tox testing.

The base OS shouldn't matter in this case either.  It's the version of python inside the container(s) that matters?  Sorry, I haven't looked at the ceph-volume tests before so I'm a little behind.

**Comment by djgalloway** [comment]
jenkins test ceph-volume tox

**Comment by guits** [comment]
> > @djgalloway so from what I'm understanding, CentOS 8 isn't a requirement for the ceph-volume tox testing.
> 
> It's the version of python inside the container(s) that matters?

No, this is well the version of python from the hypervisor that matters

**Comment by guits** [comment]
> Umm, so as you mentioned, the builders that run these jobs are CentOS7 so they don't have python3. If these jobs require python3, then we should upgrade the builders to CentOS8.

looks like it's not always true:

```bash
[jenkins-build@smithi015 ~]$ cat /etc/redhat-release
CentOS Linux release 7.4.1708 (Core)
[jenkins-build@smithi015 ~]$ python2 --version
Python 2.7.5
[jenkins-build@smithi015 ~]$ python3 --version
Python 3.6.8
[jenkins-build@smithi015 ~]$
```

**Comment by djgalloway** [comment]
Should [this](https://github.com/ceph/ceph/pull/37274#issuecomment-790679968) suffice then?

**Comment by guits** [comment]
> Should [this](https://github.com/ceph/ceph/pull/37274#issuecomment-790679968) suffice then?

That might work

**Comment by djgalloway** [comment]
> > Should [this](https://github.com/ceph/ceph/pull/37274#issuecomment-790679968) suffice then?
> 
> That might work

https://github.com/ceph/ceph-build/pull/1766

**Comment by guits** [comment]
jenkins test ceph-volume batch centos8-bluestore-mixed_type

**Comment by guits** [comment]
jenkins test ceph-volume batch mixed_type_dmcrypt_explicit

**Comment by guits** [comment]
jenkins test ceph-volume batch centos8-bluestore-single_type

**Comment by guits** [comment]
jenkins test ceph-volume batch centos8-filestore-single_type

**Comment by guits** [comment]
jenkins test ceph-volume lvm centos8-bluestore-create

**Comment by guits** [comment]
jenkins test ceph-volume lvm centos8-filestore-create

**Comment by guits** [comment]
jenkins test ceph-volume lvm centos8-bluestore-dmcrypt

**Comment by guits** [comment]
jenkins test ceph-volume batch centos8-bluestore-mixed_type_explicit

**Comment by guits** [comment]
jenkins test ceph-volume batch centos8-bluestore-single_type_dmcrypt

**Comment by guits** [comment]
jenkins test ceph-volume batch centos8-filestore-mixed_type_explicit

**Pull request closed without merging**